CMakeLists.txt

```diff
@@ -0,0 +1,160 @@
+
+option(FREERTOS_KERNEL_USE_CUSTOM_PORT "Enable the use of a custom port implementation from a different project. When enabling this option set <FREERTOS_KERNEL_PORT> to the path that contains port files." OFF)
+option(FREERTOS_KERNEL_USE_CUSTOM_HEAP "Enable the use of a custom heap implementation from a different project. When enabling this option set <FREERTOS_KERNEL_HEAP_FILE> to the file that implements the memory management." OFF)
+
+## FreeRTOS Kernel Config ##
+# Check that <FREERTOS_KERNEL_CONFIG> contains FreeRTOSConfig.h
+if(NOT DEFINED FREERTOS_KERNEL_CONFIG)
+    message(FATAL_ERROR "<FREERTOS_KERNEL_CONFIG> is undefined. Set <FREERTOS_KERNEL_CONFIG> to the path that contains FreeRTOSConfig.h.")
+else()
+    if(NOT EXISTS "${FREERTOS_KERNEL_CONFIG}/FreeRTOSConfig.h")
+        message(FATAL_ERROR "FreeRTOSConfig.h not found in ${FREERTOS_KERNEL_CONFIG}. Set <FREERTOS_KERNEL_CONFIG> to the path that contains FreeRTOSConfig.h.")
+    endif()
+endif ()
+
+## FreRTOS Kernel Port ##
+if(FREERTOS_KERNEL_USE_CUSTOM_PORT)
+    # If a custom port is being used then check that <FREERTOS_KERNEL_PORT> contains a valid directory.
+    if(NOT DEFINED FREERTOS_KERNEL_PORT)
+        message(FATAL_ERROR "<FREERTOS_KERNEL_PORT> is undefined. When using a custom port set <FREERTOS_KERNEL_PORT> to the path that contains the port source files.")
+    else()
+        if(NOT EXISTS "${FREERTOS_KERNEL_PORT}")
+            message(FATAL_ERROR "Custom port folder: ${FREERTOS_KERNEL_PORT} does not exist.")
+        endif()
+    endif()
+else()
+    # If using a default port then check the compiler ID identified by CMake and set the port folder accordingly.
+    if(${CMAKE_C_COMPILER_ID} STREQUAL "ARMClang")
+        set(FREERTOS_COMPILER_PORT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/portable/ARMClang")
+    elseif(${CMAKE_C_COMPILER_ID} STREQUAL "Bruce")
+        set(FREERTOS_COMPILER_PORT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/portable/BCC/16BitDOS")
+    elseif(${CMAKE_C_COMPILER_ID} STREQUAL "GNU")
+        set(FREERTOS_COMPILER_PORT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/portable/GCC")
+    elseif(${CMAKE_C_COMPILER_ID} STREQUAL "IAR")
+        set(FREERTOS_COMPILER_PORT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/portable/IAR")
+    elseif(${CMAKE_C_COMPILER_ID} STREQUAL "MSVC")
+        set(FREERTOS_COMPILER_PORT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/portable/MSVC-MingW")
+    elseif(${CMAKE_C_COMPILER_ID} STREQUAL "OpenWatcom")
+        set(FREERTOS_COMPILER_PORT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/portable/oWatcom/16BitDOS")
+    elseif(${CMAKE_C_COMPILER_ID} STREQUAL "SDCC")
+        set(FREERTOS_COMPILER_PORT_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/portable/SDCC")
+    else()
+        message(FATAL_ERROR "The C compiler identification is ${CMAKE_C_COMPILER_ID}. No port is available for this compiler. Check your toolchain configuration or set <FREERTOS_KERNEL_USE_CUSTOM_PORT> to use a custom FreeRTOS port.")
```

Comment:
Any reason for leaving out the other ports? Or are these to be added when needed?

Reply:
`portable/MSVC-MingW` x86 port is not only for MSVC, but for GCC (MinGW64) too.
With your current code, it is not possible to build x86 port with GCC (Windows or Linux) so far.

Reply:
And it doesn't handle ports in `portable/ThirdParty/GCC`.

Reply:
>I tired to match up all the ports for compilers identified by CMake listed here

Thanks for sharing your justification. Totally makes sense. I'd probably do the same thing if I were in your shoes.
I do think the ports your PR handles could be expanded without significant trouble so I won't hold up the PR for it.

Reply:
> `portable/MSVC-MingW` x86 port is not only for MSVC, but for GCC (MinGW-w64) too.
This is a good point and something I didn't think about. It wouldn't be impossible because the user could manually provide the port path, but it would be misleading and something they shouldn't expect to have to do for a natively supported port.

I'll think about how this could be improved.